caput/pipeline.py

```diff
@@ -605,45 +643,164 @@ def run(self):
-                    receiving_task._pipeline_inspect_queue_product(task._out_keys, out)
+
+            if self.execution_order == "legacy":
+                # Advance the task pointer. If this is the last task
+                # in the list, it will get wrapped when selecting
+                # the next task
+                self._task_idx += 1
+
+            # Ensure the output(s) are correctly structured
+            out = self._check_task_output(out, task)
+
+            if out is None:
+                continue
+
+            # Queue outputs for any associated tasks
+            for key, product in zip(task._out_keys, out):
+                # Try to pass this product to each task
+                received = [
+                    recv._pipeline_queue_product(key, product)
+                    for recv in self.tasks
+                    if recv is not None
+                ]
+
+                if not any(received):
+                    # Just warn. This probably shouldn't happen, but there
+                    # could be some edge cases to deal with
+                    warnings.warn(
+                        f"Task {task!s} tried to pass key {key} "
+                        "but no task was found to accept it."
+                    )
+
+        # Pipeline is done
+        logger.info("FIN")
+
+    def _next_task(self):
+        """Get the next task to run from the task list.
+
+        Task is chosen based the following criteria:
+        - able to do something in its current state
+        - highest priority
+        - highest base priority
+        - next in pipeline config order
+
+        If no task is available to do anything, restart from the
+        pipeline task head.
+        """
+        # Get a list of tasks which are availble to run
+        available = []
+
+        for ii in range(len(self.tasks)):
+            # Loop through tasks starting at the current index. Including
+            # the current tasks first ensures we clear out completed
+            # tasks faster
+            jj = (ii + self._task_idx) % len(self.tasks)
+            task = self.tasks[jj]
+
+            if task is None:
+                continue
+
+            if task._pipeline_is_available:
+                available.append(jj)
+
+        if not available:
+            # Nothing is currently available, so fall back to a
+            # blind loop starting at the first task. If there is
+            # nothing left, this will raise StopIteration.
+            self._task_idx = self._task_head
+            return self._iter_tasks()
+
+        # Reverse sort the available tasks first by priority and second
+        # by base priority such that for any two tasks with equal priority,
+        # the task with highest base priority will be selected
+        new_index = sorted(
+            available,
+            key=lambda i: (self.tasks[i].priority, self.tasks[i].base_priority),
+            reverse=True,
+        )[0]
```

Comment:
I thought about it briefly, but evenly for a pipeline with a couple hundred tasks I think the extra overhead should be negligible, so I generally would favour readability over optimization

Reply:
The other thing to note here is that it's only sorting the *available* tasks rather than the entire task list, and generally only a handful of tasks will be available at any given time

Reply:
I think it's best to keep the code as simple as you have it now, at least until the optimisation becomes important.